One to One Track Association

gawebb-dstl opened the pull request:
I've implemented a different track to track association algorthim using the 'assign2D' method in stonesoup.dataassociator._assignment. This track to track association only allows tracks to be associated on a one to one basis, so tracks can't be associated to multiple other tracks. 

The algorithm itself is pretty arbitary and can/should be improved; it just looks back over the last 10 states in a track and applies a distance measure to those states

nperree-dstl commented on stonesoup/measures.py:
```diff
@@ -8,6 +8,58 @@
+class GenericMeasure(Base):
+    """Measure base type
+
+    A measure provides a means to assess the seperation between two
+     objects item1 and item2.
```
Space causes something weird with formatting...
```suggestion
    A measure provides a means to assess the seperation between two
    objects item1 and item2.
```

nperree-dstl commented on stonesoup/measures.py:
```diff
@@ -8,6 +8,58 @@
+class TrackMeasure(GenericMeasure):
+    """Measure base type
+
+    A measure provides a means to assess the seperation between two
+     objects item1 and item2.
```
Same issue as above
```suggestion
    A measure provides a means to assess the separation between two
    objects item1 and item2.
```

nperree-dstl commented on stonesoup/measures.py:
```diff
@@ -8,6 +8,58 @@
+        return NotImplementedError
+
+
+class TrackMeasure(GenericMeasure):
```
What is the need for TrackMeasure as well as GenericMeasure, they look the same to me?

gawebb-dstl replied:
They are the same really. TrackMeasure is just an abstract class to it make clear which measure is needed for track association

nperree-dstl reviewed (changes requested):
Mostly small commments as looks fine to me but could do with more explanation on the example and some documentation added for the associator.

orosoman-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,179 @@
+
+class OneToOneAssociatorWithGates(Associator):
+
+    gates: List[GeneralAssociationGate] = Property(default=None)
+
+    fail_gate_value: Union[int, float, complex, np.number] = Property(default=None)
+
+    measure: GenericMeasure = Property()
+    association_threshold: float = Property(default=None)
+
+    maximise_measure: bool = Property(default=False)
```
All of these  (as well as the class) could do with some documentation.  This is important so the user doesn't have to go through the source code to work out how to use it.  There's plenty of good examples of how to write this if you have a browse around the source files.  If you based this on a paper, a reference to that would be handy as well.

gawebb-dstl replied:
There wasn't enough documentation (hopefully there is now). This isn't based on a paper.

orosoman-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,179 @@
+
+class OneToOneAssociatorWithGates(Associator):
+
+    gates: List[GeneralAssociationGate] = Property(default=None)
+
+    fail_gate_value: Union[int, float, complex, np.number] = Property(default=None)
+
+    measure: GenericMeasure = Property()
+    association_threshold: float = Property(default=None)
+
+    maximise_measure: bool = Property(default=False)
```
All of these  (as well as the class) could do with some documentation.  This is important so the user doesn't have to go through the source code to work out how to use it.  There's plenty of good examples of how to write this if you have a browse around the source files.  If you based this on a paper, a reference to that would be handy as well.

orosoman-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,179 @@
+
+    maximise_measure: bool = Property(default=False)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.gates is None:
+            self.gates = []
+
+        if self.fail_gate_value is None:
+            if self.maximise_measure:
+                self.fail_gate_value = 0
+            else:
+                self.fail_gate_value = 1e6
+
+        if self.association_threshold is None:
+            if self.maximise_measure:
+                self.association_threshold = 0
+            else:
+                self.association_threshold = 1e6
```
This `__init__` isn't necessary - Stone Soup handles for you.  If you set the default for `gates` to `[]` and the other two here to `None`, it would be a lot cleaner.  Will just require a few changes/extra if statements later on, but setting `association_threshold` to arbitrarily large/small numbers is probably best avoided

gawebb-dstl replied:
It's my understanding that when you set a default to a mutable object such as a list, all instances of that class point to the same object. You can see from the example below:

```
from stonesoup.base import Base, Property

class A(Base):
    var: list = Property(default=[])

a1 = A()
a1.var.append(2)
a2 = A()
print(a2.var)
#prints [2]
```

So if you use a default that is mutable you need `__init__` . At that point I thought I might as well add the other bits of code

I wanted to make the gates in this class optional so that meant using default values. I agree it's best practice not to use arbitrarily large/small numbers but I think the example failed with I tried to use infinity as the default for the `fail_gate_value`

orosoman-dstl replied:
Good point about mutable objects.  It probably makes sense to have an `__init__` for that, but why not default `association_threshold` to `None` and later on you could for example enclose lines 144-151 in an `if self.association_threshold:` statement.

orosoman-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,179 @@
+
+    maximise_measure: bool = Property(default=False)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.gates is None:
+            self.gates = []
+
+        if self.fail_gate_value is None:
+            if self.maximise_measure:
+                self.fail_gate_value = 0
+            else:
+                self.fail_gate_value = 1e6
+
+        if self.association_threshold is None:
+            if self.maximise_measure:
+                self.association_threshold = 0
+            else:
+                self.association_threshold = 1e6
```
This `__init__` isn't necessary - Stone Soup handles for you.  If you set the default for `gates` to `[]` and the other two here to `None`, it would be a lot cleaner.  Will just require a few changes/extra if statements later on, but setting `association_threshold` to arbitrarily large/small numbers is probably best avoided

gawebb-dstl commented:
Some notes on the recent changes

- I've removed gating from the `OneToOneAssociator` class. It simplifies the class. Gating can be added later if needed by subclassing and overriding the `individual_weighting` function or by creating a custom measure. 
- I've added more documentation (hopefully enough). Let me know if there needs to be any more. 
- I've moved non-state measures into a separate file. I would like to move the measures into a folder and then have multiple files in that folder, however that would be a breaking change which I'd like to avoid.
- The `OneToOneAssociator` class was created to associate general objects. It uses the Hungarian algorithm to optimise this 1 to 1 pairing. I feel like there is a gap for another association method (e.g. greedy nearest neighbour) for general objects.

nperree-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,162 @@
+from typing import Tuple, Collection
```
This file needs adding to docs/source/stonesoup.dataassociator.rst so that modules appear in the docs

gawebb-dstl replied:
Good spot

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,354 @@
+"""
+
+# %%
+# First some generic imports and variables are set up as they are needed for all of the examples
```
```suggestion
# First some generic imports and variables are set up as they are needed for all of the examples.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,354 @@
+plotter.fig
+
+# %%
+# It’s not immediately clear which states should be associated to each other
```
```suggestion
# It’s not immediately clear which states should be associated to each other.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,354 @@
+
+# %%
+# The `track_measure` will take the multiple measures from `full_state_sequence_measure` and
+# condense it down into one single measure by taking the mean
```
```suggestion
# condense it down into one single measure by taking the mean.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,354 @@
+# Generic Association Example
+# -----------------------------------
+# The association algorithm can be to associated many things. In this example we'll associate words
+# to demonstrate the versatility of the algorithm
```
```suggestion
# to demonstrate the versatility of the algorithm.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,354 @@
+
+# %%
+# An example scenario where an application can only accept standard colour names. However the
+# input parameter can accept CCS colours
```
```suggestion
# input parameter can accept CCS colours.
```

nperree-dstl reviewed (changes requested):
A couple of minor comments to address and some missing full stops in the example.

mharris-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,167 @@
+
+    def associate(self, objects_a: Collection, objects_b: Collection) \
+            -> Tuple[AssociationSet, Collection, Collection]:
+        """Associate two collections of objects together.
```
How? There's a comment in the source code explaining but not easiest to see.

mharris-dstl commented on stonesoup/dataassociator/general.py:
```diff
@@ -0,0 +1,167 @@
+
+    def association_dict(self, objects_a: Collection, objects_b: Collection) -> dict:
+        """
+        This is a wrapper function around the `associate` function. The two collections of objects
```
```suggestion
        This is a wrapper function around :func:`associate`. The two collections of objects
```
Or something similar if that doesn't work.

mharris-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,169 @@
+    """
+
+    def __call__(self, collection_1: Collection, collection_2: Collection) -> float:
+        r"""
```
```suggestion
        """
```

mharris-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,169 @@
+    def __call__(self, collection_1: Collection, collection_2: Collection) -> float:
+        r"""
+        The measure is calculated by finding the number of items in common between the two
+        collections and divides it by the total number of unique items in the combined collection
```
```suggestion
        collections and divides it by the total number of unique items in the combined collection.
```

mharris-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,169 @@
+import heapq
```
Looking at the docs, a lot of these are quite messy when rendered. Would it be possibly to make some changes to make them look a bit cleaner/add missing descriptions for parameters?

Quite a few full stops need adding in as well.

mharris-dstl reviewed (changes requested):
Hopefully these are useful comments. Happy to discuss if you disagree with anything :)

gawebb-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,358 @@
+"""
+Generic One-to-One Association Examples
```
Other 1-2-1 association algorithms can only associate tracks or states. This is generic because it can associate anything with a valid measure function.

gawebb-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,358 @@
+# **Track a3** is too far away to associated with any of the other tracks.
+#
+# **Track b3** appears to be closer to `Track a1` than `Track b1` is. However, `Track b3` takes
+# place 20 seconds after `Track a1` and `Track b1`, therefore there are no overlapping time
```
The plot just highlights that in a 2D plot there may be tracks that look similar but they actually aren't. The tracks were designed to highlight that issue.

gawebb-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,358 @@
+received_colours_scheme = ["FloralWhite", "LightGreen", "Magenta"]
+
+# %%
+# The association process
```
I've explained it in a previous section now and don't want to repeat it

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -62,7 +67,7 @@
 
 # %%
-# This scenario has been created so it’s not immediately clear which states should be associated
+# This scenario has been created, so it’s not immediately clear which states should be associated
```
```suggestion
# This scenario has been designed such the optimal association between ``states_from_a`` and ``states_from_b`` is unclear to the human eye.
```
The added comma here makes the phrasing confusing. The suggestion may need to go over a couple of lines.

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -71,25 +76,27 @@
-# distance between pairs of objects. For pairs of objects with a measure equal to or above the
-# :attr:`~.OneToOneAssociator.threshold`, these pairs won’t be associated together.
+# **Associate States**. The :class:`~.OneToOneAssociator` will minimise the total measure
+# (:class:`~.Euclidean` distance) between the two states. The ``OneToOneAssociator`` uses scipy's
```
```suggestion
# (:class:`~.Euclidean` distance) between the two states. The ``OneToOneAssociator`` uses SciPy's
```

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -316,28 +337,35 @@ def __call__(self, word_1: str, word_2: str) -> float:
 # ^^^^^^^^^^^^^^^^^^^^^^^
 # ``BetterWordMeasure`` looks for words that are identical or is a word/phrase is contained within
```
```suggestion
# ``MatchingWordMeasure`` looks for words that are identical or is a word/phrase is contained within
```

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -316,28 +337,35 @@ def __call__(self, word_1: str, word_2: str) -> float:
 
 
 # %%
 # The association process
```
```suggestion
# The association process:
```

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -316,28 +337,35 @@ def __call__(self, word_1: str, word_2: str) -> float:
 association_dict = associator.association_dict(standard_colours, received_colours_scheme)
 
+# %%
+# Print association results
```
```suggestion
# Print association results:
```

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -352,7 +380,18 @@ def __call__(self, word_1: str, word_2: str) -> float:
 association_dict = associator.association_dict(standard_colours, received_colours_scheme)
 
+# %%
+# Print association results
```
```suggestion
# Print association results:
```

mharris-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -352,7 +380,18 @@ def __call__(self, word_1: str, word_2: str) -> float:
+# The :class:`~.OneToOneAssociator` can be used for multiple varied purposes. It was created
+# originally for track association but can be used to associate anything. The examples above show
+# its use in associating :class:`~.StateMutualSequence`, :class:`~.State` and :class:`str`.
+# It’s a flexible association class that can be tailored for the scenario.
```
```suggestion
# It’s a flexible association class that can be tailored for many use cases.
```

mharris-dstl reviewed (changes requested):
A few bits to change.

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+
+# %%
+# **Associate States**. The :class:`~.OneToOneAssociator` will minimise the total measure
+# (:class:`~.Euclidean` distance) between the two states. The ``OneToOneAssociator`` uses SciPy's
```
```suggestion
# (:class:`~.Euclidean` distance) between the two states. The `~.OneToOneAssociator` uses SciPy's
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+plotter.fig
+
+# %%
+# This scenario has been designed such the optimal association between ``states_from_a`` 
+# and ``states_from_b`` is unclear to the human eye.
```
```suggestion
# This scenario has been designed such the optimal association between `states_from_a` 
# and `states_from_b` is unclear to the human eye.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+# ^^^^^^^^^^^^^^^^
+
+# %%
+# We have states from source A and source B marked as ``states_from_a`` and ``states_from_b``
```
```suggestion
# We have states from source A and source B marked as `states_from_a` and `states_from_b`
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+tracks_b = {track_b1, track_b2, track_b3}
+
+# %%
+# The tracks are plotted. As before, we use different colours to separate ``tracks_a`` from
+# ``tracks_b``.
```
```suggestion
# The tracks are plotted. As before, we use different colours to separate `tracks_a` from
# `tracks_b`.
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+# %%
+# The ``full_state_sequence_measure`` (:class:`~.StateSequenceMeasure`) measure will apply the
+# Euclidean state measure to each state in the tracks, with the same time. This produces a multiple
+# measures for each state.
+full_state_sequence_measure = StateSequenceMeasure(state_measure=Euclidean(mapping=[0, 1]))
+
+# %%
+# ``track_measure`` (:class:`~.MeanMeasure`) will take the multiple measures from
+# ``full_state_sequence_measure`` and condense it down into one single measure by taking the mean.
+track_measure = MeanMeasure(measure=full_state_sequence_measure)
```
```suggestion
# The `full_state_sequence_measure` (:class:`~.StateSequenceMeasure`) measure will apply the
# Euclidean state measure to each state in the tracks, with the same time. This produces a multiple
# measures for each state.
full_state_sequence_measure = StateSequenceMeasure(state_measure=Euclidean(mapping=[0, 1]))

# %%
# `track_measure` (:class:`~.MeanMeasure`) will take the multiple measures from
# `full_state_sequence_measure` and condense it down into one single measure by taking the mean.
track_measure = MeanMeasure(measure=full_state_sequence_measure)
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+# %%
+# Measure 1 - Example 1
+# ^^^^^^^^^^^^^^^^^^^^^^^
+# ``WordMeasure`` is a crude measure to compare how similar words are. It calculates the
```
```suggestion
# :class:`WordMeasure` is a crude measure to compare how similar words are. It calculates the
```

nperree-dstl commented on docs/examples/One_to_One_Associator.py:
```diff
@@ -0,0 +1,397 @@
+# %%
+# Measure 2 - Example 1
+# ^^^^^^^^^^^^^^^^^^^^^^^
+# ``MatchingWordMeasure`` looks for words that are identical or is a word/phrase is contained within
```
```suggestion
# :class:`MatchingWordMeasure` looks for words that are identical or is a word/phrase is contained within
```

nperree-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,183 @@
+class MultipleMeasure(BaseMeasure):
+    """ MultipleMeasure base class.
+
+    This measure produces a list of ``float`` values instead of a singular ``float`` value. This
```
```suggestion
    This measure produces a list of `float` values instead of a singular `float` value. This
```

nperree-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,183 @@
+class TrackMeasure(BaseMeasure):
+    """TrackMeasure base class.
+
+    A measure provides a means to assess the separation between two track objects ``track_1``
+    and ``track_2``. It should return a float value of the distance measure between the two tracks.
```
```suggestion
    A measure provides a means to assess the separation between two track objects `track_1`
    and `track_2`. It should return a float value of the distance measure between the two tracks.
```

nperree-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,183 @@
+                 state_sequence_2: StateMutableSequence,
+                 times_to_measure: Sequence[datetime] = None) -> List[float]:
+        """
+        Compare the states from each state sequence for every time in ``times_to_measure``.
+
+        If ``times_to_measure`` is None. Find all times that both state sequences have in common.
+
+        Parameters
+        ----------
+        state_sequence_1 : :class:`.~StateMutableSequence`
+            a state sequence to compare against ``state_sequence_2``.
+        state_sequence_2 : :class:`.~StateMutableSequence`
+            a state sequence to compare against ``state_sequence_1``.
+        times_to_measure : Sequence of :class:`.~datetime`
+            Calculate the state measure for states in the state sequences at these times. Default
+            value is ``None``. If ``None``, ``times_to_measure`` is calculated as all the times
+            that both state sequences have in common.
```
```suggestion
        Compare the states from each state sequence for every time in `times_to_measure`.

        If `times_to_measure` is None. Find all times that both state sequences have in common.

        Parameters
        ----------
        state_sequence_1 : :class:`.~StateMutableSequence`
            a state sequence to compare against `state_sequence_2`.
        state_sequence_2 : :class:`.~StateMutableSequence`
            a state sequence to compare against `state_sequence_1`.
        times_to_measure : Sequence of :class:`.~datetime`
            Calculate the state measure for states in the state sequences at these times. Default
            value is `None`. If `None`, `times_to_measure` is calculated as all the times
            that both state sequences have in common.
```

nperree-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,183 @@
+class RecentStateSequenceMeasure(MultipleMeasure):
+    """
+    Applies a state measure to each state in the state sequence with for the most recent *n*
+    matching times. It will return less than ``n_states_to_compare`` values if there are less
```
```suggestion
    matching times. It will return less than `n_states_to_compare` values if there are less
```

nperree-dstl commented on stonesoup/non_state_measures.py:
```diff
@@ -0,0 +1,183 @@
+    def __call__(self, state_sequence_1: StateMutableSequence,
+                 state_sequence_2: StateMutableSequence) -> List[float]:
+        """
+        Compare the states from each state sequence for the most recent ``n_states_to_compare``
+        times.
+
+        Parameters
+        ----------
+        state_sequence_1 : :class:`.~StateMutableSequence`
+            a state sequence to compare against ``state_sequence_2``.
+        state_sequence_2 : :class:`.~StateMutableSequence`
+            a state sequence to compare against ``state_sequence_1``.
+
+        Returns
```
```suggestion
        Compare the states from each state sequence for the most recent `n_states_to_compare`
        times.

        Parameters
        ----------
        state_sequence_1 : :class:`.~StateMutableSequence`
            a state sequence to compare against `state_sequence_2`.
        state_sequence_2 : :class:`.~StateMutableSequence`
            a state sequence to compare against `state_sequence_1`.

        Returns
```

nperree-dstl reviewed (approved):
A few more documentation comments. I found the variable names being in red in the example/documentation a bit off-putting but if it was intentional feel free to ignore me.

Otherwise looks good!

sdhiscocks commented:
I think it's worth creating a measures as it's own module with submodules. You could move `stonesoup/measures.py` to `stonesoup/measures/__init__.py` and then move and rename `non_state_measures.py` into the `measures` directory. May want to break the new file to different names, so like `stonesoup.measures.multi` and `stonesoup.measures.summary`?
